strutil/env.go

```diff
@@ -0,0 +1,253 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+)
+
+// RawEnvironment is a slice of key=value strings.
+//
+// Using this type is required for low-level interactions, like invoking
+// execve. In other cases using Environment is both sufficient and safer.
+type RawEnvironment []string
+
+// String returns the readable, quoted representation of the raw environment.
+func (raw RawEnvironment) String() string {
+	return Quoted(raw)
+}
+
+// Environment is an unordered map of key=value strings.
```

Comment:
os represents Environment as []string  so probably Enviroment itself should be an OrderedMap as well

Reply:
I started with an ordered map but discarded that later on as it simplifies a few things. Order is irrelevant when there are no duplicates. Environment is specifically not RawEnvironment, with all of the associated complexity of duplicates, malformed entries and the like. When we produce a RawEnvironment we always sort entries so the results are deterministic where it matters.

Reply:
what does it simplifies? just syntax? a map is a map in the end

Reply:
I wanted RawEnvironment to be deterministic irrespective of the order of construction of the environment. Currently I produce order sorted by keys, which is natural for an otherwise unordered map.

When Environment had an intrinsic order discarding it felt wrong. At the same time the order of environment entries should really not be anything but an implementation detail so having order there was weird.

I looked back to when app and snap environment overrides were introduced. The only reason we had order was to support expanding expressions. Since Environment specifically has just key=values, not key=expressions, having order was not required.

I only kept order for EnvironmentDelta because we already represented environment overrides (in the yaml) as ordered.

Reply:
> I wanted RawEnvironment to be deterministic irrespective of the order of construction of the environment.

I understand this but it's again quite a change from current behavior of snapd, somebody can have a test expecting overridden things at the tail of what they get, and now they will be sorted.

Reply:
I'm not sure if I understand your concern correctly but environment order is not something applications perceive through most APIs. Only low level code can observe the order. As long as there are no maliciously crafted entries (not key=value, not duplicated keys) the order is totally irrelevant.

Reply:
Specifically after we chain through snap-exec the resulting order is not even the order we set in snap-run. It is the order as it was transformed by libc running in snap-confine.

Reply:
I looked a bit more and snapenv itself scrambles everything because of the through map roundtrip there, so my point here is moot. OTOH we should try to add as little as possible new code for the new types.

Reply:
I think everything in this tread is addresed.